File: src/days/day1.rs
use super::Solve;

pub struct Day1;
impl Solve<i128> for Day1 {
    fn solve_part_1(&mut self, input: &str) -> i128 {
        let mut left = vec![];
        let mut right = vec![];
        Self::parsery(input, &mut left, &mut right);
        left.iter()
            .zip(right.iter())
            .map(|(a, b)| a.abs_diff(*b))
            .sum::<u128>() as i128
    }

    fn new() -> Self {
        Self
    }

    fn solve_part_2(&mut self, input: &str) -> i128 {
        let mut left = vec![];
        let mut right = vec![];
        Self::parsery(input, &mut left, &mut right);
        left.iter()
            .map(|i| i * right.iter().filter(|j| *i == **j).count() as i128)
            .sum()
    }
}

impl Day1 {
    fn parsery(input: &str, left: &mut Vec<i128>, right: &mut Vec<i128>) {
        input
            .lines()
            .map(|i| i.split_whitespace())
            .for_each(|mut i| {
                left.push(i.next().unwrap().parse::<i128>().unwrap());
                right.push(i.next().unwrap().parse::<i128>().unwrap());
            });
        left.sort();
        right.sort();
    }
}

File: aoc_cpp/day1.cpp
#include <algorithm>
import derived;
import utils;
export module day1;
class Day1: public Solution<long long int> {
    public:
        std::tuple<std::vector<long long int>, std::vector<long long int>> parse(std::string input) {
            auto left = std::vector<long long int>();
            auto right = std::vector<long long int>();
            std::stringstream stream(input);
            std::string line;
            for (std::string line: split(input)) {
                auto splitted = split(line, ' ');
                left.push_back(std::stoi(splitted[0]));
                right.push_back(std::stoi(splitted.back()));
            }
            std::sort(left.begin(), left.end());
            std::sort(right.begin(), right.end());
            return std::make_tuple(left, right);
        }
        
        long long int solve_part_1(std::string input) {
            auto [left, right] = parse(input);
            auto sum = 0;
            for (auto i = 0; i < left.size(); i++) {
                sum += std::abs(left[i] - right[i]);
            }
            return sum;
        }

        long long int solve_part_2(std::string input) {
            auto [left, right] = parse(input);
            auto count = 0;
            for (auto i: left) {
                auto filtered = 0;
                for (auto j: right) {
                    if (i == j) {
                        filtered += 1;
                    }
                }
                count += i * filtered;
            }
            return count;
        }
};
